### PR TITLE
Specify comparison semantics for identifiers.

### LINKED ISSUE #110: Define and document the comparison semantics for identity names and keys.
Current engine implementations use case-sensitive (or more specifically, byte-wise) comparisons. This is probably the only sane approach; otherwise the spec needs to define case-folding and normalisation behavior which I'd rather avoid if possible. 

The spec, however, currently nominates that these identifiers must be UTF-8 strings and I'm wondering now if this is a mistake and should instead be treated as opaque byte sequences.

Either way, the recommendation about UUIDs stands, and needs to be clarified to indicate that they should be provided as lower-case, hypen-separated, hex-encoded ASCII strings (whatever this format is called), for example: `1bff04c5-3fc5-4d18-be94-313529c57b99`.

@ezzatron I would be keen to get your opinion about this at some point.

### PR DESCRIPTION
**Pull request by jmalloc**
Fixes #110.

This PR introduces a documentation/specification change that clarifies how identifiers (names and keys) are compared. The decisions are recorded in ADR-12, included in this PR.

### CODE REVIEW

File: aggregate.go [hunk 1]
```diff
@@ -108,13 +108,18 @@ type AggregateConfigurer interface {
...
-	// Both the name and the key MUST be non-empty UTF-8 strings consisting
-	// solely of printable Unicode characters, excluding whitespace. A printable
-	// character is any character from the Letter, Mark, Number, Punctuation or
-	// Symbol categories.
+	// key MUST NOT be changed. The RECOMMENDED key format is an RFC 4122 UUID
+	// represented as a hyphen-separated, lowercase hexadecimal string, such as
+	// "5195fe85-eb3f-4121-84b0-be72cbc5722f".
+	//
+	// Both identifiers MUST be non-empty UTF-8 strings consisting solely of
+	// printable Unicode characters, excluding whitespace. A printable character
```
Comment (danilvpetrov) [inline aggregate.go]: Is whitespace actually a printable character?

Comment (jmalloc) [reply aggregate.go]: Yeah, as weird it sounds, whitespace is in the set of characters called "printable". It's distinguished from things like control characters.

Comment (danilvpetrov) [reply aggregate.go]: Cool, just wanted to specify, thanks.

File: aggregate.go [hunk 2]
```diff
@@ -108,13 +108,18 @@ type AggregateConfigurer interface {
...
+	// is any character from the Letter, Mark, Number, Punctuation or Symbol
```
Comment (danilvpetrov) [inline aggregate.go]: Should we specify that this terms (i.e. `Letter`, `Mark`, `Number`, `Punctuation` or `Symbol`) are used golang `unicode` package to disambiguate the definition?

File: docs/adr/0012-identifier-comparison.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+of the source code. They do not need to be parsed and validated from user input.
+
+## Decision
+
+In keeping with current behavior, we've decided to specify byte-wise comparison
+semantics for identifiers.
+
+## Consequences
+
+Existing tooling and engine implementations, such as `configkit` and the the
```
Comment (koden-km) [inline docs/adr/0012-identifier-comparison.md]:
```suggestion
Existing tooling and engine implementations, such as `configkit` and the
```

File: aggregate.go (hunk 2 above)
Comment (jmalloc) [reply aggregate.go]: I don't think we need to explain that. Those categories are defined by the [Unicode specification](https://unicodebook.readthedocs.io/unicode.html#categories) itself, not Go.

File: docs/adr/0012-identifier-comparison.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+in-memory engine implementation in `testkit` do not require changed. Identifiers
```
[the remark this answers is not in the document]
Comment (jmalloc) [reply docs/adr/0012-identifier-comparison.md]:
```suggestion
in-memory engine implementation in `testkit` do not require changes. Identifiers
```

[the remark this answers is not in the document]
Comment (jmalloc) [reply docs/adr/0012-identifier-comparison.md]: I reworded this slightly.

[the remark this answers is not in the document]
Comment (koden-km) [reply docs/adr/0012-identifier-comparison.md]: So Go's standard comparision operators use byte-wise compares?

[the remark this answers is not in the document]
Comment (jmalloc) [reply docs/adr/0012-identifier-comparison.md]: Yes, as Go's `string` type can contain any sequence of bytes (not just UTF-8).  If you want to do case-folding/normalization you have to write that yourself, as there's no single approach that's applicable in all circumstances.

**Comment by jmalloc** [comment]
Thanks guys.

**Pull request merged**